Retry-After header handling fixes

ldipenti opened the pull request:
## Retry-After header handling fixes

### Description

* Pass response headers to exception_from_message() when receiving an error response.
* Pass Retry-After header to BaseHTTPException, some other errors than 429 may include it. (Example: 503 Service Unavailable)
* Retry-After header may include an HTTP-date value, so check if that's the case and if so, translate it to a delay seconds value.
* Added tests.

### Status

- done, ready for review

### Checklist (tick everything that applies)

- [X] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [ ] Documentation
- [X] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [X] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

ldipenti committed:
Pass response headers to exception_from_message() when receiving an
error response.
Pass Retry-After header to BaseHTTPException, some other errors than
429 may include it. (Example: 503 Service Unavailable)
Retry-After header may include an HTTP-date value, so check if that's
the case and if so, translate it to a delay seconds value.
Added tests.

tetron commented on libcloud/common/exceptions.py:
```diff
@@ -47,7 +51,7 @@ class RateLimitReachedError(BaseHTTPError):
     message = '%s Rate limit exceeded' % (code)
 
     def __init__(self, *args, **kwargs):
```
Can we make sure that `code`, `message` and `headers` attributes are set on `RateLimitReachedError` ?  The fact that `RateLimitReachedError` extends `BaseHTTPError` but doesn't set these fields has caused real problems for us in our exception handlers in the past.

ldipenti replied:
Sure, will add a superclass constructor call to it.

tetron commented on libcloud/common/exceptions.py:
```diff
@@ -68,7 +72,11 @@ def exception_from_message(code, message, headers=None):
-    if headers and 'retry_after' in headers:
-        kwargs['retry_after'] = headers['retry_after']
+    if headers and 'retry-after' in headers:
+        http_date = parsedate_tz(headers['retry-after'])
```
Have you confirmed this returns `None` on parse error instead of throwing an exception?

ldipenti replied:
Yes, as described on the docs, parsedate_tz() returns None or a tuple.

tetron commented on libcloud/common/exceptions.py:
```diff
@@ -68,7 +72,11 @@ def exception_from_message(code, message, headers=None):
+        if http_date is not None:
+            # Convert HTTP-date to delay-seconds
+            delay = max(0, int(mktime_tz(http_date) - time.time()))
+            headers['retry-after'] = str(delay)
```
This header munging behavior should be documented somewhere, starting with the documentation string for the function.

ldipenti replied:
Updated docstring with examples.

tetron replied:
Thanks!

ldipenti committed:
RateLimitReachedError exception sets message, code and headers.
Added docstring documentation describing HTTP-date format
translation on Retry-After header.

tetron commented on libcloud/common/exceptions.py:
```diff
@@ -47,7 +51,14 @@ class RateLimitReachedError(BaseHTTPError):
+        headers = kwargs.pop('headers', None)
+        super(RateLimitReachedError, self).__init__(self.code,
+                                                    self.message,
+                                                    headers)
```
This doesn't look right, I don't think `self.code` and `self.message` are bound yet?  That's the point of calling the superclass constructor...

tetron replied:
And if that's really a bug, the test coverage needs to catch it.

tetron replied:
I see now, they are class variables, so that's okay.  Nevermind.

tetron commented:
@ldipenti this LGTM

pquentin commented:
@ldipenti Great! But there's still the RateLimitReachedError issue mentioned by @tetron?

ldipenti commented:
@pquentin class vars are accessible by __init__() and are passed to the superclass' constructor, right? Maybe I'm missing something.

pquentin commented:
Oh, right, class vars, sorry.

ldipenti commented:
@pquentin do you have any other suggestions for this PR?

pquentin commented:
Not really, it sounds good! I'll probably merge it tomorrow or Friday unless someone does it first. 

Unfortunately the process we have means merging takes some work, with signoffs and updates to CHANGES.rst. I also like to use fixups when rebasing in order to minimize the number of commits that don't build, which helps with bisecting.

ldipenti commented:
Cool, thanks!

pquentin commented:
Thanks! Merged in trunk. ✨
